File: js/mcap.js
/** First-class support for the frozen synapse/1 MCAP profile.
 *
 * Container encoding and decoding are delegated to @mcap/core. This module
 * only applies the Synapse profile, catalog, embedded BFBS, and fixed-struct
 * wrapper.
 */
import { McapStreamReader, McapWriter } from '@mcap/core';
import {
  mcapMessageEncoding,
  mcapMetadataName,
  mcapProfile,
  mcapSchemaEncoding,
  mcapSchemaSetHashKey,
  mcapSessionIdKey,
  mcapSourceKey,
  mcapTimeBasisCorrelated,
  mcapTimeBasisKey,
  mcapTimeBasisMonotonicBoot,
  mcapTimeBasisUnixEpoch,
  mcapTopicIdKey,
  schemaSetHash,
  topicByName,
} from './topic_catalog.js';

export * as container from '@mcap/core';

export const TimeBasis = Object.freeze({
  MONOTONIC_BOOT: mcapTimeBasisMonotonicBoot,
  UNIX_EPOCH: mcapTimeBasisUnixEpoch,
  CORRELATED: mcapTimeBasisCorrelated,
});

function validSessionId(value) {
  return /^[0-9a-f]{32}$/.test(value);
}

function resolveTopic(value) {
  if (typeof value !== 'string') return value;
  const topic = topicByName(value);
  if (!topic) throw new Error(`unknown Synapse topic: ${value}`);
  return topic;
}

/** Load a package BFBS in Node or a browser/bundler deployment. */
export async function loadSchema(topic) {
  const info = resolveTopic(topic);
  const url = new URL(info.mcapSchemaFile, import.meta.url);
  if (url.protocol === 'file:') {
    const { readFile } = await import('node:fs/promises');
    return new Uint8Array(await readFile(url));
  }
  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`failed to load ${url}: ${response.status} ${response.statusText}`);
  }
  return new Uint8Array(await response.arrayBuffer());
}

/** Wrap fixed struct bytes in their existing one-field FlatBuffer root. */
export function wrapFixedPayload(payload) {
  const data = payload instanceof Uint8Array ? payload : new Uint8Array(payload);
  const objectSize = data.byteLength + 4;
  if (objectSize > 0xffff || data.byteLength > 0xffffffff - 14) {
    throw new Error('fixed payload is too large to wrap');
  }
  const output = new Uint8Array(data.byteLength + 14);
  const view = new DataView(output.buffer);
  view.setUint32(0, 4, true);
  view.setInt32(4, -objectSize, true);
  output.set(data, 8);
  view.setUint16(8 + data.byteLength, 6, true);
  view.setUint16(10 + data.byteLength, objectSize, true);
  view.setUint16(12 + data.byteLength, 4, true);
  return output;
}

/** Uncompressed, unchunked, index-less synapse/1 writer. */
export class Writer {
  constructor({
    writable,
    library,
    sessionId,
    source,
    timeBasis = TimeBasis.MONOTONIC_BOOT,
    schemaLoader = loadSchema,
  }) {
    if (!library) throw new Error('MCAP library identifier is empty');
    if (!source) throw new Error('Synapse MCAP source is empty');
    if (!validSessionId(sessionId)) {
      throw new Error('Synapse MCAP session id must be 32 lowercase hexadecimal characters');
    }
    if (!Object.values(TimeBasis).includes(timeBasis)) {
      throw new Error(`unsupported Synapse MCAP time basis: ${timeBasis}`);
    }
    this._inner = new McapWriter({
      writable,
      useStatistics: false,
      useSummaryOffsets: false,
      useChunks: false,
      repeatSchemas: false,
      repeatChannels: false,
      useAttachmentIndex: false,
      useMetadataIndex: false,
      useMessageIndex: false,
      useChunkIndex: false,
    });
    this._library = library;
    this._metadata = new Map([
      [mcapSchemaSetHashKey, schemaSetHash],
      [mcapSessionIdKey, sessionId],
      [mcapSourceKey, source],
      [mcapTimeBasisKey, timeBasis],
    ]);
    this._schemaLoader = schemaLoader;
    this._schemaIds = new Map();
    this._started = false;
  }

  async start() {
    if (this._started) throw new Error('Synapse MCAP writer is already started');
    await this._inner.start({ profile: mcapProfile, library: this._library });
    await this._inner.addMetadata({ name: mcapMetadataName, metadata: this._metadata });
    this._started = true;
  }

  async addTopic(topic, channelTopic) {
    if (!this._started) throw new Error('call start() before addTopic()');
    const info = resolveTopic(topic);
    let schemaId = this._schemaIds.get(info.id);
    if (schemaId === undefined) {
      schemaId = await this._inner.registerSchema({
        name: info.mcapSchemaName,
        encoding: mcapSchemaEncoding,
        data: await this._schemaLoader(info),
      });
      this._schemaIds.set(info.id, schemaId);
    }
    const id = await this._inner.registerChannel({
      schemaId,
      topic: channelTopic,
      messageEncoding: mcapMessageEncoding,
      metadata: new Map([[mcapTopicIdKey, String(info.id)]]),
    });
    return { id, topic: info, nextSequence: 0 };
  }

  async write(channel, logTimeNs, publishTimeNs, data) {
    const bytes = data instanceof Uint8Array ? data : new Uint8Array(data);
    await this._inner.addMessage({
      channelId: channel.id,
      sequence: channel.nextSequence,
      logTime: BigInt(logTimeNs),
      publishTime: BigInt(publishTimeNs),
      data: bytes,
    });
    channel.nextSequence = (channel.nextSequence + 1) >>> 0;
  }

  async writeFixed(channel, logTimeNs, publishTimeNs, payload) {
    const bytes = payload instanceof Uint8Array ? payload : new Uint8Array(payload);
    if (!channel.topic.fixedLayout) {
      throw new Error(`${channel.topic.name} is not fixed-layout`);
    }
    if (bytes.byteLength !== channel.topic.payloadSize) {
      throw new Error(
        `fixed payload is ${bytes.byteLength} bytes, expected ${channel.topic.payloadSize}`,
      );
    }
    await this.write(channel, logTimeNs, publishTimeNs, wrapFixedPayload(bytes));
  }

  async finish() {
    if (!this._started) throw new Error('call start() before finish()');
    await this._inner.end();
  }
}

/** Streaming reader that validates the frozen Synapse profile as records arrive. */
export class Reader {
  constructor(options) {
    this._inner = new McapStreamReader(options);
    this._headerSeen = false;
    this._metadataSeen = false;
  }

  append(data) {
    this._inner.append(data);
  }

  nextRecord() {
    const record = this._inner.nextRecord();
    if (!record) return undefined;
    if (record.type === 'Header') {
      if (record.profile !== mcapProfile) {
        throw new Error(`unsupported MCAP profile: ${record.profile}`);
      }
      this._headerSeen = true;
    } else if (record.type === 'Metadata' && record.name === mcapMetadataName) {
      if (!/^[0-9a-f]{32}$/.test(record.metadata.get(mcapSchemaSetHashKey) ?? '')) {
        throw new Error('invalid Synapse MCAP schema-set hash');
      }
      if (!validSessionId(record.metadata.get(mcapSessionIdKey) ?? '')) {
        throw new Error('invalid Synapse MCAP session id');
      }
      if (!(record.metadata.get(mcapSourceKey) ?? '')) {
        throw new Error('missing Synapse MCAP source');
      }
      if (!Object.values(TimeBasis).includes(record.metadata.get(mcapTimeBasisKey))) {
        throw new Error('invalid Synapse MCAP time basis');
      }
      this._metadataSeen = true;
    } else if (record.type === 'Message' && (!this._headerSeen || !this._metadataSeen)) {
      throw new Error('message precedes required Synapse MCAP header or metadata');
    }
    return record;
  }

  done() {
    return this._inner.done();
  }

  assertValid() {
    if (!this.done()) throw new Error('MCAP input is incomplete');
    if (!this._headerSeen) throw new Error('missing Synapse MCAP header');
    if (!this._metadataSeen) throw new Error('missing required Synapse MCAP metadata');
  }
}

File: rust/src/mcap.rs
//! Optional host-side implementation of the frozen `synapse/1` MCAP profile.
//!
//! Enable the crate's `mcap` feature. The embedded C writer in the C release
//! archive follows the same profile and is tested against this implementation.

use std::{collections::BTreeMap, io::Seek, io::Write};

use ::mcap::records::{MessageHeader, Metadata};

use crate::{
    schemas::schema_by_name,
    topic_catalog::{
        MCAP_MESSAGE_ENCODING, MCAP_METADATA_NAME, MCAP_PROFILE, MCAP_SCHEMA_ENCODING,
        MCAP_SCHEMA_SET_HASH_KEY, MCAP_SESSION_ID_KEY, MCAP_SOURCE_KEY, MCAP_TIME_BASIS_CORRELATED,
        MCAP_TIME_BASIS_KEY, MCAP_TIME_BASIS_MONOTONIC_BOOT, MCAP_TIME_BASIS_UNIX_EPOCH,
        MCAP_TOPIC_ID_KEY, SCHEMA_SET_HASH, TopicInfo,
    },
};

/// The timestamp basis recorded in required `synapse/1` metadata.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TimeBasis {
    MonotonicBoot,
    UnixEpoch,
    Correlated,
}

impl TimeBasis {
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::MonotonicBoot => MCAP_TIME_BASIS_MONOTONIC_BOOT,
            Self::UnixEpoch => MCAP_TIME_BASIS_UNIX_EPOCH,
            Self::Correlated => MCAP_TIME_BASIS_CORRELATED,
        }
    }
}

/// A registered channel and its caller-owned sequence state.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct TopicChannel {
    id: u16,
    topic_id: u16,
    next_sequence: u32,
    payload_size: Option<usize>,
    fixed_layout: bool,
}

impl TopicChannel {
    pub const fn id(&self) -> u16 {
        self.id
    }
}

#[derive(Debug)]
pub enum Error {
    Container(::mcap::McapError),
    EmptyLibrary,
    EmptySource,
    InvalidSessionId,
    MissingSchema(&'static str),
    NotFixedLayout(&'static str),
    PayloadSize { expected: usize, actual: usize },
    ChannelTopicMismatch,
    PayloadTooLarge,
}

impl std::fmt::Display for Error {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Container(error) => error.fmt(formatter),
            Self::EmptyLibrary => formatter.write_str("MCAP library identifier is empty"),
            Self::EmptySource => formatter.write_str("Synapse MCAP source is empty"),
            Self::InvalidSessionId => formatter
                .write_str("Synapse MCAP session id must be 32 lowercase hexadecimal characters"),
            Self::MissingSchema(file) => {
                write!(formatter, "embedded BFBS is missing for {file}")
            }
            Self::NotFixedLayout(topic) => write!(formatter, "{topic} is not fixed-layout"),
            Self::PayloadSize { expected, actual } => {
                write!(
                    formatter,
                    "fixed payload is {actual} bytes, expected {expected}"
                )
            }
            Self::ChannelTopicMismatch => {
                formatter.write_str("fixed payload type does not match the MCAP channel")
            }
            Self::PayloadTooLarge => formatter.write_str("fixed payload is too large to wrap"),
        }
    }
}

impl std::error::Error for Error {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            Self::Container(error) => Some(error),
            _ => None,
        }
    }
}

impl From<::mcap::McapError> for Error {
    fn from(error: ::mcap::McapError) -> Self {
        Self::Container(error)
    }
}

pub type Result<T> = std::result::Result<T, Error>;

/// Implemented by every generated fixed-layout Synapse topic struct.
pub trait FixedMessage {
    fn topic() -> &'static TopicInfo;
    fn as_payload_bytes(&self) -> &[u8];
}

/// Uncompressed, unchunked, index-less host writer for `synapse/1` logs.
pub struct Writer<W: Write + Seek> {
    inner: ::mcap::Writer<W>,
    fixed_buffer: Vec<u8>,
}

impl<W: Write + Seek> Writer<W> {
    pub fn new(
        output: W,
        library: &str,
        session_id: &str,
        source: &str,
        time_basis: TimeBasis,
    ) -> Result<Self> {
        if library.is_empty() {
            return Err(Error::EmptyLibrary);
        }
        if source.is_empty() {
            return Err(Error::EmptySource);
        }
        if !is_lowercase_hex_128(session_id) {
            return Err(Error::InvalidSessionId);
        }

        let options = ::mcap::WriteOptions::new()
            .profile(MCAP_PROFILE)
            .library(library)
            .compression(None)
            .use_chunks(false)
            .emit_summary_records(false)
            .emit_summary_offsets(false)
            .emit_message_indexes(false)
            .calculate_data_section_crc(false)
            .calculate_summary_section_crc(false);
        let mut inner = options.create(output)?;
        let metadata = BTreeMap::from([
            (
                MCAP_SCHEMA_SET_HASH_KEY.to_owned(),
                SCHEMA_SET_HASH.to_owned(),
            ),
            (MCAP_SESSION_ID_KEY.to_owned(), session_id.to_owned()),
            (MCAP_SOURCE_KEY.to_owned(), source.to_owned()),
            (
                MCAP_TIME_BASIS_KEY.to_owned(),
                time_basis.as_str().to_owned(),
            ),
        ]);
        inner.write_metadata(&Metadata {
            name: MCAP_METADATA_NAME.to_owned(),
            metadata,
        })?;
        Ok(Self {
            inner,
            fixed_buffer: Vec::new(),
        })
    }

    /// Register one selected topic before writing its first message.
    pub fn add_topic(
        &mut self,
        topic: &'static TopicInfo,
        channel_topic: &str,
    ) -> Result<TopicChannel> {
        let schema =
            schema_by_name(topic.schema_file).ok_or(Error::MissingSchema(topic.schema_file))?;
        let schema_id =
            self.inner
                .add_schema(topic.mcap_schema_name, MCAP_SCHEMA_ENCODING, schema.bfbs)?;
        let metadata = BTreeMap::from([(MCAP_TOPIC_ID_KEY.to_owned(), topic.id.to_string())]);
        let id =
            self.inner
                .add_channel(schema_id, channel_topic, MCAP_MESSAGE_ENCODING, &metadata)?;
        Ok(TopicChannel {
            id,
            topic_id: topic.id,
            next_sequence: 0,
            payload_size: topic.payload_size,
            fixed_layout: topic.fixed_layout,
        })
    }

    pub fn add_fixed_topic<T: FixedMessage>(
        &mut self,
        channel_topic: &str,
    ) -> Result<TopicChannel> {
        let topic = T::topic();
        if !topic.fixed_layout {
            return Err(Error::NotFixedLayout(topic.name));
        }
        self.add_topic(topic, channel_topic)
    }

    /// Write an already table-wrapped topic payload.
    pub fn write(
        &mut self,
        channel: &mut TopicChannel,
        log_time_ns: u64,
        publish_time_ns: u64,
        data: &[u8],
    ) -> Result<()> {
        let sequence = channel.next_sequence;
        channel.next_sequence = channel.next_sequence.wrapping_add(1);
        self.inner.write_to_known_channel(
            &MessageHeader {
                channel_id: channel.id,
                sequence,
                log_time: log_time_ns,
                publish_time: publish_time_ns,
            },
            data,
        )?;
        Ok(())
    }

    /// Wrap a generated fixed-layout struct and write it as its existing root
    /// table. The host scratch vector is retained and reused across messages.
    pub fn write_fixed<T: FixedMessage>(
        &mut self,
        channel: &mut TopicChannel,
        log_time_ns: u64,
        publish_time_ns: u64,
        payload: &T,
    ) -> Result<()> {
        if channel.topic_id != T::topic().id {
            return Err(Error::ChannelTopicMismatch);
        }
        if !channel.fixed_layout {
            return Err(Error::NotFixedLayout(T::topic().name));
        }
        let bytes = payload.as_payload_bytes();
        let expected = channel.payload_size.unwrap_or(0);
        if bytes.len() != expected {
            return Err(Error::PayloadSize {
                expected,
                actual: bytes.len(),
            });
        }
        wrap_fixed_payload(&mut self.fixed_buffer, bytes)?;
        let sequence = channel.next_sequence;
        channel.next_sequence = channel.next_sequence.wrapping_add(1);
        self.inner.write_to_known_channel(
            &MessageHeader {
                channel_id: channel.id,
                sequence,
                log_time: log_time_ns,
                publish_time: publish_time_ns,
            },
            &self.fixed_buffer,
        )?;
        Ok(())
    }

    pub fn flush(&mut self) -> Result<()> {
        self.inner.flush()?;
        Ok(())
    }

    pub fn finish(mut self) -> Result<W> {
        self.inner.finish()?;
        Ok(self.inner.into_inner())
    }
}

fn is_lowercase_hex_128(value: &str) -> bool {
    value.len() == 32
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

fn wrap_fixed_payload(output: &mut Vec<u8>, payload: &[u8]) -> Result<()> {
    let object_size = payload.len().checked_add(4).ok_or(Error::PayloadTooLarge)?;
    if object_size > u16::MAX as usize || payload.len() > u32::MAX as usize - 14 {
        return Err(Error::PayloadTooLarge);
    }
    output.clear();
    output.reserve(payload.len() + 14);
    output.extend_from_slice(&4_u32.to_le_bytes());
    output.extend_from_slice(&(-(object_size as i32)).to_le_bytes());
    output.extend_from_slice(payload);
    output.extend_from_slice(&6_u16.to_le_bytes());
    output.extend_from_slice(&(object_size as u16).to_le_bytes());
    output.extend_from_slice(&4_u16.to_le_bytes());
    Ok(())
}

/// The conformant upstream container implementation remains available for
/// dynamic reading, recovery, and advanced host-only operations.
pub use ::mcap as container;
